[nathan] 스프링 카페 2단계 - 글 쓰기 기능 구현

nathan29849 opened the pull request:
안녕하세요 나단입니다.
스프링 카페 2단계 PR을 보냅니다.

우선 리뷰어 분들께 항상 감사드립니다.  

1단계에서 PR 리뷰 받은 부분에 대해 많이 생각해보고 이번 미션 2단계에 임하였습니다.
그럼에도 불구하고, 스프링이 처음이라 부족한 부분이 많을 것이라 생각합니다.
이번 단계의 리뷰에서도 많이 배워나가겠습니다.

다시 한 번 감사드립니다.

---

## 참고 사항 

- 이번에는 리드미에 나름 보기 편할 수 있게 tree 정보와 view 이미지를 각 기능별로 첨부하였습니다.
- Controller 계층에 대한 테스트 코드는 작성하지 못했습니다.
- 수동 빈 등록에서 자동 빈 등록으로 바꿨습니다. (AppConfig 제거 - 실질적으로 사용하지 않았기 때문)
- 기존에 존재했던 `UserForm` DTO를 그대로 update시에도 사용할까 하였으나, 새로운 비밀번호 등 다른 부분이 존재한다고 판단이 들어 새로운 DTO인 `UpdateUserForm`을 만들어 사용하였습니다.

## 질문 사항
- 미션 추가 요구사항을 구현할 때, `@PutMapping`을 이용하면서, HTTP Method인 PUT과 PATCH의 차이점에 대해서 학습을 하게 되었는데,  제가 이해한 바로는 다음과 같습니다. (멱등성에 대해서는 제대로 이해하지 못했습니다. 다시 학습이 필요할 것 같습니다.)
    - PUT : 자원의 전체적인 교체 시 이용됨. 멱등성 지킴
    - PATCH : 자원의 일부 교체 시 이용됨.  멱등성 안지킴
    - 이번 미션에서는 아이디, 이름, 이메일, 비밀번호를 모두 `UpdateUserForm`이라는 DTO에 담아 전달했기 때문에 PUT을 쓰는게 맞다고 생각했습니다. 허나 보통의 웹사이트(ex. naver)에서는 비밀번호 변경은 따로 존재합니다. 이 때는 PATCH를 쓰는 것이 맞는 것인지 여쭤보고 싶었습니다.
   
- 이번에도 1단계와 같이 테스트 코드 작성시, `Service`와 `Repository` 계층에 있는 public 메서드만 테스트하는 코드를 작성했습니다.
   - 좀 더 디테일(세분화?)하게 테스트 코드를 작성해야하는 것인지 여쭙고자 합니다. 
    
---

다시 한 번 감사드립니다.

nathan29849 committed:
[step2] refactor : qna 관련 html 중복 제거

thymeleaf 문법을 통한 중복 제거

nathan29849 committed:
[step2] feat : 게시글 작성 기능 추가

ArticleController, ArticleService, ArticleRepository 생성
Article, ArticleForm(DTO) 생성

nathan29849 committed:
[step2] mod : HomeController 글 목록 나타내게 수정

HomeController 수정
subnav 부분 수정하여 해당항목일 때 focus 되도록 정함

nathan29849 committed:
[step2] refactor : id 값을 모두 Long에서 int로 변경

nathan29849 committed:
[step2] mod : 게시글 상세보기 - view 수정

qna/show.html 수정

nathan29849 committed:
[step1] refactor : 1단계 PR에 대한 리팩토링 진행

- Service 계층의 인터페이스를 제거(지나친 추상화 방지)
- UserForm을 통해 Service 계층에서 User를 생성할 수 있도록 함(domain은 Service와 직결)
- Optional 값을 받아올 때, get() 말고, orElseThrow() 사용
- id 값을 지우고, index로 유저 서치 (ArrayList의 장점-탐색을 이용하기 위함)
- synchronizedList 사용 (동시성 문제를 해결하기 위해)

nathan29849 committed:
[nathan29849] add : gitignore MacOS 정보 추가

nathan29849 committed:
[step2] refactor : 1단계 PR 리뷰 내용을 2단계 적용

- Article domain은 Service 내에서만 처리하도록 수정
- id를 없애고 index로 사용(ArrayList 장점 이용-탐색)
- ArticleService 인터페이스 제거 - 지나친 추상화 막기(YAGNI)
- synchronizedList 사용(동시성 문제 해결)

nathan29849 committed:
[step2] refactor : ArrayList의 인덱스+1로 url 매핑

- User, Article 인스턴스 변수인 index 값을 set할 때, +1만큼 받고,
  세부 사항(유저 프로필 또는 게시글 show)을 들어가는 url의 index 값을 컨트롤러에서 읽을 때,
  index-1로 받아냄
- qna/show.html 부분 수정(subnavbar 안보이는 현상 수정)

nathan29849 committed:
[step2] add : ArticleService Test 코드 추가

ArticleService의 기능들을 테스트할 수 있는 테스트 코드 추가

nathan29849 committed:
[step2] del : AppConfig 삭제 - 스프링빈 자동등록

- @Service, @Repository 애노테이션을 이용하여 스프링 빈 자동등록

nathan29849 committed:
[step2] add : MemoryArticleRepository TEST 작성

- MemoryArticleRepository의 public 함수들(save(), findByIndex(), findAll())에 대해 테스트

nathan29849 committed:
[step2] feat : 유저 수정 기능 구현

- application property 추가 (hiddenmethod filter)
- 유저 수정 정보를 담은 DTO인 UpdateUserForm 추가
- updateForm.html 추가
- Controller, Service, Repository 계층에서 기능 구현에 필요한 메서드 추가

nathan29849 committed:
[step2] add : 유저 수정 기능에 대한 테스트 작성

- UserService에 대한 Test 작성
- MemoryUserRepository에 대한 테스트 작성

nathan29849 committed:
[step2] docs : README - 스프링카페 2단계

- Tree 추가
- 2단계 학습 목표 및 기능 요구사항 추가
- 2단계 View 추가(게시글, 유저수정)

wheejuni commented on src/main/java/com/kakao/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,37 @@
+
+    @GetMapping("articles/{index}")
+    public String detail(@PathVariable("index") int index, Model model) {
+        ArticleForm articleForm = articleService.findOneArticle(index-1);
```
`PathVariable` 의 인덱스 표현과 실제 DB상의 인덱스 값이 시작하는 지점이 다르다면 잠재적으로 버그 가능성이 너무 많습니다~
통일하시는 것이 좋기는 합니다.

nathan29849 replied:
미션 2단계에서
 `ArrayList에서 index - 1 해당하는 데이터를 조회한 후 Model에 저장해 /qna/show.html에 전달한다.` 라는 요구사항이 있어 이렇게 구현했습니다.
 만약 통일 하려고 하면 더미 데이터를 인덱스 0에 넣어주고, 조회나 변경이 불가능하게 하면 될까요?

wheejuni commented on src/main/java/com/kakao/cafe/domain/Article.java:
```diff
@@ -0,0 +1,47 @@
+        this.index = index+1;
+    }
+
+    public boolean compareBy(int index) {
+        return this.index == index;
```
`.equals()` 오버라이드로 해결할 순 없는 건가요?

nathan29849 replied:
아 그럴 수 있겠네요! 다시 코드를 확인해보니 쓰이는 부분이 없어서 지워도 될 것 같습니다!

wheejuni commented on src/main/java/com/kakao/cafe/repository/MemoryArticleRepository.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public Article save(Article article) {
+        for (int i = 0; i < articleStore.size(); i++) {
+            if (articleStore.get(i) == null) {
+                return store(article, i);
+            }
+        }
```
헐~ 
이 로직의 시간 복잡도가 어떻게 되는걸까요? 나중에 게시물이 10만개, 20만개 들어오면요? 물론 극단적인 가정이긴 합니다만...

nathan29849 replied:
미션에서 `ArrayList`를 사용하라는 요구사항이 있었습니다.
이를 고려하여, `Article` 객체의 삽입이 일어날 때, 무조건 `add`로 추가하여 자료형의 크기를 무분별하게 늘리기 보다는, 
순차탐색을 통해 이전에 `Article` 객체가 삭제된 인덱스를 찾아내어 집어 넣으려고 했습니다.

말씀하신대로, 해당 로직의 시간 복잡도가 엄청 커질 우려가 있다라는 점에  동의합니다.
사실 이를 작성하면서 indentation depth도 깊어지고, 더 좋은 방법이 있을 것 같았지만, 방법을 찾지는 못했습니다.
지금 간단히 생각나는 것으로는 `minheap`이나 `queue` 자료구조를 활용하여 인덱스만 따로 관리하도록 하는 방법이 떠올랐습니다.

wheejuni commented on src/main/java/com/kakao/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,33 @@
+        this.articleRepository = articleRepository;
+    }
+
+    public String post(ArticleForm articleForm) {
```
제목을 리턴한다? 좀 어색한데요.... 받아쓰는 측에서 리턴받은 객체에게 제목을 물어보면 되지 않으려나요?

nathan29849 replied:
사실 post를 테스트할 때 사용하려고 String 타입으로 반환을 받았습니다.
말씀하신대로 받아쓰는 측에서 객체에게 제목을 물어볼 수 있기도 하고,
post에 대한 테스트가 꼭 리턴값이 있어야 가능한 것이 아니라는 생각이 듭니다.

wheejuni reviewed (changes requested):
수고 많으셨습니다 💯 
방향성에 크게 문제는 없고요, 참고하실만한 코멘트 몇 개 드렸는데 참고 및 수정 부탁드립니다~ 👍 
질문은 별도의 댓글로 답변 드릴게요

wheejuni commented:
- 테스트 코드에 관하여 
**공개된 API를 테스트** 하는 방향성으로 가시면 됩니다. 정확히 말하면... 구현과 테스트는 스펙을 향해야 하고, 메서드는 스펙에 대한 하나의 추상화에 불과한 거긴 한데요, 만약 테스트 대상이 되는 클래스로 노출하는 API가 `public` 한 메서드로 다 표현되는 스펙을 갖고 있다면... 퍼블릭 메서드에 대한 테스트 케이스만 모두 구현되어도 테스트는 충분히 된 것이라고 봅니다.

- `PUT` `PATCH` 에 대하여
사실 아마 네이버는 계정 서비스에 대해서 get, post말곤 쓰는게 없을 겁니다(...)
이해하신 방향은 정확한 것 같고요, **비밀번호** 라는 단일한 관심사에 대한 전체 변경이라는 차원에서 저라면 비번 변경이여도 PUT을 쓸 것 같네요.
사실 유저 입장에서 네이버가 관리하는 회원정보가 어떤 것들인지 잘 알 수 없고(법리적이거나 사회공학적인 부분이 아닌, 순수 UI 및 서비스 도메인 상을 말씀드리는 겁니다) 그걸 다 알 필요도 없기는 합니다. 즉, 뭐 철학적인 얘기일 수도 있겠지만 유저는 **유저 정보** 라는 자원을 변경한 게 아니라 _로그인 비밀번호_ 라는 자원의 전체 값을 변경한 액션을 수행한 거라고 봅니다. 

PATCH의 경우 클라이언트가 전체 자원 목록을 정확히 인지하고 있고 분명하게 그 중 일부를 변경하는 액션을 취하는 경우(예를 들어, REST API 기반으로 통신하는 검색 엔진 클러스터의 색인 기준 변경) 정도에 사용하면 되지 않을까 생각해 봅니다.

nathan29849 commented:
PR 및 질문사항에 대한 답변 모두 잘 보았습니다!
정말 감사드립니다!!

---

### 추가적으로 궁금한 점 1

PUT과 PATCH에 대해서 답변해주신 부분(`사실 유저 입장에서 네이버가 관리하는 회원정보가 어떤 것들인지 잘 알 수 없고`)에서, 
궁금한 점이 하나 있는데요

유저입장에서는 회원정보를 모두 다 알지 못한다는 말은 이해가 됐습니다.
다만, 그런 부분을 UI에서 각각 보내도록 '항목마다 개별 버튼'을 만든다면(예를들어 `비밀번호 수정하기` 같은 특정 항목에 대한 버튼)
PATCH를 사용해도 무방한 것일까요?
(순전히 개인적인 궁금증에서 질문을 드려봅니다.. - 사실 답변 주신 것만으로도 많은 도움이 됐습니다!)
 
---

### 추가적으로 궁금한 점 2

혹시 제가 리뷰 달린 부분을 모두 수정하고 다시 push를 해야  merge를 해주시는 것인지 궁금합니다!

---

다시 한 번 주말에 귀한 시간 내셔서 리뷰해주신 점 감사드립니다!!

nathan29849 committed:
[step2] refactor : 미션2 리뷰기반 1차 리팩토링 진행

- MemoryArticleRepository : articleStore에final 추가
- Article : compareBy 메서드 삭제 (미사용)
- ArticleService : post의 반환 타입을 void로 변경(굳이 제목을 반환할 필요가 없다고 판단)

nathan29849 committed:
[step3] init : h2, jdbc 초기 설정

- build.gradle 초기 설정
- properties 초기 설정
- sql/schema.sql : ARTICLE 테이블 생성 sql 추가
